KsApi/extensions/NSURLSession.swift

```diff
@@ -12,6 +12,65 @@ private func parseJSONData(_ data: Data) -> Any? {
 private let scheduler = QueueScheduler(qos: .background, name: "com.kickstarter.ksapi", targeting: nil)
 
 internal extension URLSession {
+  // Wrap an URLSession producer with Graph error envelope logic.
+  func rac_graphDataResponse(_ request: URLRequest)
+    -> SignalProducer<Data, GraphError> {
+    let producer = self.reactive.data(with: request)
+
+    return producer
+      .start(on: scheduler)
+      .flatMapError { error -> SignalProducer<(Data, URLResponse), GraphError> in
+        print("🔴 [KsApi] Request Error \(error.localizedDescription)")
+
+        return .init(error: GraphError.requestError(error, nil))
+      }
+      .flatMap(.concat) { data, response -> SignalProducer<Data, GraphError> in
+        guard let response = response as? HTTPURLResponse else { fatalError() }
+
+        guard
+          (200..<300).contains(response.statusCode),
+          let headers = response.allHeaderFields as? [String: String],
+          let contentType = headers["Content-Type"], contentType.hasPrefix("application/json")
+        else {
+          print("🔴 [KsApi] HTTP Failure \(self.sanitized(request))")
+
+          let producer = SignalProducer<Data, GraphError>(
+            error: .jsonDecodingError(
+              responseString: String(data: data, encoding: .utf8),
+              error: nil
+            )
+          )
+
+          return self.decodeGraphErrorsOrCoalesce(from: data, coalesce: producer)
+        }
+
+        print("🔵 [KsApi] HTTP Success \(self.sanitized(request))")
+
+        let producer = SignalProducer<Data, GraphError>(value: data)
+
+        return self.decodeGraphErrorsOrCoalesce(from: data, coalesce: producer)
+      }
+  }
+
+  private func decodeGraphErrorsOrCoalesce(
```

Comment:
Lol I know I kind've pieced this together when we paired, but I think this would be better as just:

```swift
private func tryDecodeGraphErrors(from data: Data) -> SignalProducer<Data, GraphError> {
  let decodedErrors = try? JSONDecoder().decode(GraphResponseErrors.self, from: data)

  guard let error = decodedErrors?.errors?.first else { return .init(value: data) }

  print("🔴 [KsApi] Graph Error \(error.message)")

  return .init(error: GraphError.decodeError(error))
}
```
^ So if it successfully decodes an error envelope is will error out the stream, else it'll just return the data that it received satisfying the type requirements.

The call sites would then look like:
```swift

return self.tryDecodeGraphErrors(from: data)
  .flatMap { data in
    SignalProducer<Data, GraphError>(
      error: .jsonDecodingError(
        responseString: String(data: data, encoding: .utf8),
        error: nil
      )
    )
  }
```
and:
```swift
return self.tryDecodeGraphErrors(from: data)
  .flatMap { data in
    SignalProducer<Data, GraphError>(value: data)
  }
```
I also don't think we need to use the `.concat` strategy on these inner producers if the encompassing producer is already using it 🤔 .